File: c++/src/vm/lowerer.cpp
#ifndef LOWERER_CPP
#define LOWERER_CPP

#include "lowerer.hpp"

InstructionLowerer::InstructionLowerer(std::vector<InstructionIr> ir_insts)
{
    this->ir_insts = ir_insts;
    return;
}

/*
Returns free temp register, and also reserves it.
*/
int InstructionLowerer::getFreeTempReg()
{
    for (auto reg_info : this->temp_regs)
    {
        if (reg_info.second == RegisterState::FREE)
        {
            this->temp_regs[reg_info.first] = RegisterState::RESERVED;
            return reg_info.first;
        }
    }
    std::cout << "No free temp register" << std::endl;
    exit(1);
}

void InstructionLowerer::freeTempReg(int reg)
{
    try
    {
        this->temp_regs[reg] = RegisterState::FREE;
    }
    catch(const std::exception& e)
    {
        std::cerr << e.what() << '\n';
        std::cout << reg << " is no temp register\n";
        exit(1);
    }
    return;
}

void InstructionLowerer::freeAllTempRegs()
{
    for (auto reg : this->temp_regs)
    {
        this->temp_regs[reg.first] = RegisterState::FREE;
    }
    return;
}

InstructionType opTokenToInstType(Token tok)
{
    if (tok.type == TokenType::PLUS)
    {
        return InstructionType::ADD;
    }
    else if (tok.type == TokenType::MINUS)
    {
        return InstructionType::SUB;
    }
    std::cout << "Cannot convert " << tokenRepr(tok) << " to InstructionType" << std::endl;
    exit(1);    
}

std::tuple<int, ArgType> tokenToArgBundle(Token tok)
{
    if (tok.type == TokenType::IMMEDIATE)
    {
        return std::make_tuple(std::stoi(tok.value), ArgType::IMMEDIATE);
    }
    else if (tok.type == TokenType::REGISTER)
    {
        return std::make_tuple(std::stoi(tok.value), ArgType::REGISTER);
    }
    else
    {
        std::cout << "Cannot convert \"" << tokenRepr(tok) << "\"" << "(" << tokenTypeRepr(tok.type) << ")" << " to inst arg";
        exit(1);
    }
}

int InstructionLowerer::binOpToInst(std::shared_ptr<BinOp> bin_op)
{
    Instruction lowerd_inst;
    lowerd_inst.type = opTokenToInstType(bin_op->op);

    bool left_is_BinOp = false;
    std::tuple<int, ArgType> left_arg_bundle;
    if (std::holds_alternative<std::shared_ptr<BinOp>>(bin_op->left))
    {
        left_is_BinOp = true;
        int left_reg = this->binOpToInst(std::get<std::shared_ptr<BinOp>>(bin_op->left));
        left_arg_bundle = {left_reg, ArgType::REGISTER};
    }
    else if (std::holds_alternative<Token>(bin_op->left))
    {
        left_arg_bundle = tokenToArgBundle(std::get<Token>(bin_op->left));
    }
    else
    {
        std::cout << "Cannot convert std::string to inst arg";
        exit(1);
    }

    bool right_is_BinOp = false;
    std::tuple<int, ArgType> right_arg_bundle;
    if (std::holds_alternative<std::shared_ptr<BinOp>>(bin_op->right))
    {
        right_is_BinOp = true;
        int right_reg = this->binOpToInst(std::get<std::shared_ptr<BinOp>>(bin_op->right));
        right_arg_bundle = {right_reg, ArgType::POINTER};
    }
    else if (std::holds_alternative<Token>(bin_op->right))
    {
        right_arg_bundle = tokenToArgBundle(std::get<Token>(bin_op->right));
    }
    else
    {
        std::cout << "Cannot convert std::string to inst arg";
        exit(1);
    }

    int ret_reg = this->getFreeTempReg();

    lowerd_inst.args[0] = ret_reg;
    lowerd_inst.arg_types[0] = ArgType::REGISTER;
    lowerd_inst.args[1] = std::get<0>(left_arg_bundle);
    lowerd_inst.arg_types[1] = std::get<1>(left_arg_bundle);
    lowerd_inst.args[2] = std::get<0>(right_arg_bundle);
    lowerd_inst.arg_types[2] = std::get<1>(right_arg_bundle);


    if (left_is_BinOp)
    {
        this->freeTempReg(std::get<0>(left_arg_bundle));
    }
    if (right_is_BinOp)
    {
        this->freeTempReg(std::get<0>(right_arg_bundle));
    }

    this->lowerd_insts.push_back(lowerd_inst);
    return ret_reg;
}


std::tuple<int, ArgType> InstructionLowerer::nodeToArgBundle(Node node)
{
    if (std::holds_alternative<std::shared_ptr<BinOp>>(node))
    {
        int ret_reg = this->binOpToInst(std::get<std::shared_ptr<BinOp>>(node));
        return std::make_tuple(ret_reg, ArgType::POINTER);
    }
    else if (std::holds_alternative<Token>(node))
    {
        return tokenToArgBundle(std::get<Token>(node));
    }
    else
    {
        std::cout << "Cannot convert std::string to inst arg";
        exit(1);
    }
}


std::vector<Instruction> InstructionLowerer::lower()
{

    for (auto inst : this->ir_insts)
    {
        this->freeAllTempRegs();
        Instruction lower_inst = {};
        lower_inst.type = inst.type;
        
        // 3 arg instructions
        if (
            (inst.type == InstructionType::ADD) ||
            (inst.type == InstructionType::SUB) ||
            (inst.type == InstructionType::MUL) ||
            (inst.type == InstructionType::IDIV) ||
            (inst.type == InstructionType::FDIV) ||
            (inst.type == InstructionType::MOD)
        )
        {
            std::tuple<int, ArgType> arg1_bundle = this->nodeToArgBundle(inst.arg1);
            std::tuple<int, ArgType> arg2_bundle = this->nodeToArgBundle(inst.arg2);
            std::tuple<int, ArgType> arg3_bundle = this->nodeToArgBundle(inst.arg3);

            lower_inst.args[0] = std::get<0>(arg1_bundle);
            lower_inst.arg_types[0] = std::get<1>(arg1_bundle);
            lower_inst.args[1] = std::get<0>(arg2_bundle);
            lower_inst.arg_types[1] = std::get<1>(arg2_bundle);
            lower_inst.args[2] = std::get<0>(arg3_bundle);
            lower_inst.arg_types[2] = std::get<1>(arg3_bundle);
        }
        // 2 arg instructions
        else if (
            (inst.type == InstructionType::MOV) ||
            (inst.type == InstructionType::ABS) ||
            (inst.type == InstructionType::CMP) ||
            (inst.type == InstructionType::LOAD) ||
            (inst.type == InstructionType::STORE) ||
            (inst.type == InstructionType::PRINT))
        {
            std::tuple<int, ArgType> arg1_bundle = this->nodeToArgBundle(inst.arg1);
            std::tuple<int, ArgType> arg2_bundle = this->nodeToArgBundle(inst.arg2);

            lower_inst.args[0] = std::get<0>(arg1_bundle);
            lower_inst.arg_types[0] = std::get<1>(arg1_bundle);
            lower_inst.args[1] = std::get<0>(arg2_bundle);
            lower_inst.arg_types[1] = std::get<1>(arg2_bundle);
        }
        // 1 arg instructions
        else if (
            (inst.type == InstructionType::PUSH) ||
            (inst.type == InstructionType::POP) ||
            (inst.type == InstructionType::DEC) ||
            (inst.type == InstructionType::INC) ||
            (inst.type == InstructionType::TIME))
        {
            std::tuple<int, ArgType> arg1_bundle = this->nodeToArgBundle(inst.arg1);

            lower_inst.args[0] = std::get<0>(arg1_bundle);
            lower_inst.arg_types[0] = std::get<1>(arg1_bundle);
        }
        // instructions with a label
        // 1 arg instructions
        else if (
            (inst.type == InstructionType::JMP) ||
            (inst.type == InstructionType::JE) ||
            (inst.type == InstructionType::JNE) ||
            (inst.type == InstructionType::JG) ||
            (inst.type == InstructionType::JL) ||
            (inst.type == InstructionType::JGE) ||
            (inst.type == InstructionType::JLE))
        {
            std::cout << this-> << std::endl;

        }


        this->lowerd_insts.push_back(lower_inst);
    }

    return this->lowerd_insts;
}



#endif

File: python/src/lowerer.py
# purely for prototyping the c++ implementation.
# it will propably not be integrated into the python version.

from src.vm_types import *


class Instruction:
    def __init__(self):
        self.op_code: InstructionType = None
        self.args: list[int] = []
        self.reg_flags: list[bool] = []

    def __repr__(self):
        return f"{self.op_code}, {self.args}, {self.reg_flags}"


class lowerer:
    def __init__(self):
        self.insts: list[Instruction] = []
        self.reg_alloc_state = {30: False, 31: False, 32: False, 33: False}

    # returns (val or reg, is_reg)
    def tokenToArg(self, token: Token) -> tuple[int, bool]:
        if token.type == TokenType.IMMEDIATE:
            return (int(token.value), False)
        if token.type == TokenType.REGISTER:
            return (int(token.value), True)
        raise Exception("Cannot convert Token to inst arg")

    def getFreeReg(self) -> int:
        for reg, is_reserved in self.reg_alloc_state.items():
            if is_reserved == False:
                self.reg_alloc_state[reg] = True
                return reg
            
        raise Exception("No free register"
)
    def freeReg(self, reg: int) -> None:
        try:
            self.reg_alloc_state[reg] = False
        except:
            raise Exception(f"r{reg} not a temp register")
        return

    def binOpToInst(self, bin_op: BinOp) -> int:
        inst = Instruction()
        inst.op_code = bin_op.op.type

        left_arg_bundle = None
        left_is_binop = False
        if bin_op.left.binop_flag == True:
            left_reg = self.binOpToInst(bin_op.left)
            left_arg_bundle = (left_reg, True)
            left_is_binop = True
        else:
            left_arg_bundle = self.tokenToArg(bin_op.left)

        right_arg_bundle = None
        right_is_binop = False
        if bin_op.right.binop_flag == True:
            right_reg = self.binOpToInst(bin_op.right)
            right_arg_bundle = (right_reg, True)
            right_is_binop = True
        else:
            right_arg_bundle = self.tokenToArg(bin_op.right)

        ret_reg = self.getFreeReg()

        inst.args.append(ret_reg)
        inst.reg_flags.append(True)
        inst.args.append(left_arg_bundle[0])
        inst.reg_flags.append(left_arg_bundle[1])
        inst.args.append(right_arg_bundle[0])
        inst.reg_flags.append(right_arg_bundle[1])
        
        if left_is_binop:
            self.freeReg(left_arg_bundle[0])
        if right_is_binop:
            self.freeReg(right_arg_bundle[0])

        self.insts.append(inst)
        return ret_reg
        

# bin = BinOp(BinOp(Token(TokenType.REGISTER, "10"), Token(TokenType.PLUS, "+"), Token(TokenType.IMMEDIATE, "2")), Token(TokenType.MINUS, "-"), Token(TokenType.REGISTER, "1"))
bin = BinOp(
    BinOp(
        BinOp(
            BinOp(
                BinOp(Token(TokenType.REGISTER, "10"), Token(TokenType.PLUS, "+"), Token(TokenType.IMMEDIATE, "2")),
                Token(TokenType.MINUS, "-"),
                Token(TokenType.REGISTER, "1"),
            ),
            Token(TokenType.PLUS, "+"),
            Token(TokenType.IMMEDIATE, "3"),
        ),
        Token(TokenType.MINUS, "-"),
        Token(TokenType.REGISTER, "4"),
    ),
    Token(TokenType.PLUS, "+"),
    BinOp(Token(TokenType.REGISTER, "5"), Token(TokenType.MINUS, "-"), Token(TokenType.IMMEDIATE, "6"))
)

lower = lowerer()
lower.binOpToInst(bin)
for inst in lower.insts:
    print(inst)
print(lower.reg_alloc_state)

